services/tasks/src/main/java/org/hyperledger/besu/services/tasks/InMemoryTasksPriorityQueues.java

```diff
@@ -0,0 +1,169 @@
+  private final Set<InMemoryTask<T>> unfinishedOutstandingTasks = new HashSet<>();
+  private final AtomicBoolean closed = new AtomicBoolean(false);
+
+  public InMemoryTasksPriorityQueues() {
+    clearInternalQueues();
+  }
+
+  private void clearInternalQueues() {
+    internalQueues.clear();
+    for (int i = 0; i < 16; i++) {
+      internalQueues.add(newEmptyQueue());
+    }
+  }
+
+  @NotNull
+  private PriorityQueue<T> newEmptyQueue() {
+    return new PriorityQueue<>(Comparator.comparingLong(TasksPriorityProvider::getPriority));
+  }
+
+  @Override
+  public synchronized void add(final T taskData) {
+    assertNotClosed();
+    var dequeue = findQueue(taskData.getDepth());
+    dequeue.add(taskData);
+  }
+
+  private PriorityQueue<T> findQueue(final int priority) {
+    while (priority + 1 > internalQueues.size()) {
+      internalQueues.add(newEmptyQueue());
+    }
+    return internalQueues.get(priority);
+  }
+
+  @Override
+  public synchronized Task<T> remove() {
+    assertNotClosed();
+    final Queue<T> lastNonEmptyQueue = findLastNonEmptyQueue();
+    if (lastNonEmptyQueue.isEmpty()) {
+      return null;
+    }
+    T data = lastNonEmptyQueue.remove();
+    InMemoryTask<T> task = new InMemoryTask<>(this, data);
+    unfinishedOutstandingTasks.add(task);
+    return task;
+  }
+
+  private Queue<T> findLastNonEmptyQueue() {
+    for (int i = internalQueues.size() - 1; i > 0; i--) {
+      final Queue<T> queue = internalQueues.get(i);
+      if (!queue.isEmpty()) {
+        return queue;
+      }
+    }
+    return internalQueues.get(0);
+  }
+
+  @Override
+  public synchronized long size() {
+    return internalQueues.stream().mapToInt(Queue::size).sum();
+  }
+
+  @Override
+  public synchronized boolean isEmpty() {
+    return findLastNonEmptyQueue().isEmpty();
+  }
+
+  @Override
+  public synchronized void clear() {
+    assertNotClosed();
+
+    unfinishedOutstandingTasks.clear();
+    clearInternalQueues();
+  }
+
+  @Override
+  public synchronized boolean allTasksCompleted() {
+    return isEmpty() && unfinishedOutstandingTasks.isEmpty();
+  }
+
+  @Override
+  public synchronized void close() {
+    closed.set(true);
+    internalQueues.clear();
+  }
+
+  private void assertNotClosed() {
+    if (closed.get()) {
+      throw new IllegalStateException("Attempt to access closed " + getClass().getSimpleName());
+    }
+  }
+
+  private synchronized void handleFailedTask(final InMemoryTask<T> task) {
+    if (markTaskCompleted(task)) {
+      add(task.getData());
+    }
+  }
+
+  private synchronized boolean markTaskCompleted(final InMemoryTask<T> task) {
+    return unfinishedOutstandingTasks.remove(task);
+  }
+
+  public synchronized boolean contains(final T request) {
+    final PriorityQueue<T> queue = findQueue(request.getDepth());
+    return queue.contains(request)
+        || unfinishedOutstandingTasks.stream()
+            .map(InMemoryTask::getData)
+            .anyMatch(data -> data == request);
```

Comment:
Interesting native `==` use there

Reply:
Do you think I should change it to something else?

Reply:
a bit unclear, but it's a smell in java. It tends to bite down the road when for some reason two objects are expected to be equal but have different pointers.

Reply:
changed to equals